File: Data Structures & Algorithms/three-integer-sum/submission-15.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) { 
        nums.sort((a,b) => a - b) 
        let res = []
        for (let i = 0; i < nums.length; i ++) {
            if (nums[i] > 0) break; 
            if (i > 0 && nums[i] === nums[i - 1]) continue

            let cur = nums[i]
            let l = i + 1 
            let r = nums.length - 1 
           
            
            while(l < r) { 
                let sum = cur + nums[l] + nums[r] 
                if (sum === 0) {
                res.push([cur, nums[l], nums[r]]) 
                while(l < r && nums[l + 1] === nums[l]) l ++ 
                while(l < r && nums[r - 1] === nums[r]) r --
                l ++ 
                r --           
            } else if (sum < 0) {
                l ++ 
            } else {
                r --
            }
            }
            
        } 

        return res
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-15.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        target = sum(nums) 
        if target % 2 != 0: 
            return False 

        target = target // 2 

        memo = {}
        def dfs(start, cur): 
            if (start, cur) in memo: 
                return memo[(start, cur)] 
            if cur == target: 
                return True 

            if cur > target: 
                return False 

            for i in range(start, len(nums)): 
                cur += nums[i] 
                if dfs(i + 1, cur): 
                    memo[(start, cur)] = True
                    return True 
                cur -= nums[i] 

            memo[(start,cur)] = False
            return False 

        return dfs(0, 0)